break Node into several separate interfaces

wjwwood opened the pull request:
This is a work in progress, I've only converted some of the most basic functionality to what I'm calling the "NodeBaseInterface". This is a preview of what I'm doing for the other parts, for which I've sketched out "NodeTopicInterface", "NodeServiceInterface", "NodeGraphInterface", and "NodeParameterInterface".

I've run into a few issues, but specifically I have some work to do to resolve the ownership issues. This is demonstrated by this:

https://github.com/ros2/rclcpp/compare/node_interfaces?expand=1#diff-14f85fa16e78532d255620f1d3529debR52

Where the comment reads "These are used to keep the scope on the containing items" and I had to comment out the Node pointer. Had we gone with inheritance, the NodeBase pointer this class would have access to would have been enough, but given the composition nature, I cannot just hold that part of the interface and maintain the same guarantee. That's ok though, I never liked how we have so much "ownership spaghetti" (for lack of a better phrase) anyways. I just have to resolve how to handle this now. I've got the options and arguments for one way versus another partially written out, I'll try to post those soon. @Karsten1987 recently ran into this in a related situation and thinks (if I can speak for him a bit) that moving away from scoped objects would help a little bit because it could ensure the lifetime of the entities created by a node do not outlive it (using a proxy or handle pattern instead).

gbiggs commented:
I think I could do with a diagram. :bowtie:

wjwwood commented:
I justed pushed the publisher and subscription parts of the refactor, `rclcpp` is building for me locally, but the examples are broken because I'll have to add some convenience functions for things like `rclcpp::spin()` which take `Node *` and extract the `NodeBaseInterface *` from it.

@gbiggs I did some design work upfront and as part of it made this:

![node_interfaces](https://cloud.githubusercontent.com/assets/100427/20777812/48754ecc-b71e-11e6-9f9d-d23757690b07.png)

I'm not sure it will end up looking exactly like that, but it will be similar.

wjwwood commented:
Things to point out in the pub/sub refactor:

- Node still has code in the `create_publisher<>()` like methods, but it's mostly variable checking.
- As much of the code that depends on the Message type or the Allocator type was put into a non-templated factory class, i.e. `PublisherFactory` and `SubscriptionFactory`.
- The non-template argument dependent code is in the `NodeTopics::create_*` functions which are also not templated and take a factory class as an argument.

---

The `Node::create_*<MessageT, ...>()` methods basically do this:

- Check arguments are valid.
- Create a `*Factory` class by calling the matching `create_*_factory<MessageT, ...>()` function.
- Pass the factory class to the `NodeTopics::create_*(..., factory, ...)` method, and get a `PublisherBase` or `SubscriptionBase` shared pointer back.
- Pass the `*Base` class to the corresponding `NodeTopics::add_*()` method.

We could consider putting the code that lives in `NodeTopics::create_*()` back into the templated `Node::create_*<MessageT, ...>()` methods, but by pushing the code into the `NodeTopics` class we minimize the code that is in the `Node::create_*<MessageT, ...>()` methods and by doing so gain:

- shorter compile times
- reduced amount of code that has to be duplicated if someone want to copy-paste-modify the Node class

Similarly, we could put the code that now lives in the factory classes (mostly code that depends on template variables in one way or another) back into the `Node::create_*<MessageT, ...>()` methods, but by putting it into a separate factory class it again improves the reusability of this code and would help whenever someone might want to copy-paste-modify the Node class. It especially reduces the amount of non-trivial, or non-boilerplate, code that needs to be copied when creating a new Node class.

----

It was not possible to have template functions on either the `NodeTopicsInterface` class or the `NodeTopics` class without also templating the `Node` class, which I wanted to avoid. That is why I didn't just move the templated code from the `Node` class into the `NodeTopics` class.

wjwwood commented:
@gbiggs just some follow up points to go with that diagram:

We wanted to accomplish a few things:

- Make it easier to mock up parts of the Node interface to more easily test the components that use them without instantiating a full Node implementation to do so.
- More clearly separate the various functionality that is aggregated by the Node class for the convenience of the user.
- Enable the creation of various other kinds of Node classes, e.g. `LifecylceNode` or `NodeWithoutParameters`.

We landed on a few things:

- Separate the current state into these discrete parts:
  - Base: core functionality and the rcl_node_t handle.
  - Topics: publisher and subscription related functionality
  - Services: service client and server related functionality
  - Parameters: parameter related functionality
  - Graph: ROS graph related functionality, i.e. wait_for_service, list_topics, etc.
- Use pure virtual "interface" classes for each of the parts, in order to make mocking them easier without worrying about side effects of the constructor.
- Use composition, rather than inheritance, to aggregate the functionality of each of parts into the Node class, i.e. `Node` has pointers to `NodeBaseInterface`, `NodeTopicsInterface`, etc. and provides forwarding methods (similar to the PIMPL pattern) rather than inheriting from them.

We discussed several of these points a lot internally and I have notes, but they're not organized at all, but I plan to put them somewhere as memoranda when I get a chance.

dirk-thomas commented on rclcpp/include/rclcpp/any_executable.hpp:
```diff
@@ -17,8 +17,13 @@
+#include "rclcpp/client.hpp"
 #include "rclcpp/macros.hpp"
-#include "rclcpp/node.hpp"
+// #include "rclcpp/node.hpp"
```
Can this be removed?

dirk-thomas commented on rclcpp/include/rclcpp/any_executable.hpp:
```diff
@@ -44,7 +49,7 @@ struct AnyExecutable
   // These are used to keep the scope on the containing items
   rclcpp::callback_group::CallbackGroup::SharedPtr callback_group;
-  rclcpp::node::Node::SharedPtr node;
+  // rclcpp::node::Node::SharedPtr node;
```
Can this be removed?

dirk-thomas commented on rclcpp/include/rclcpp/node_interfaces/node_base.hpp:
```diff
@@ -0,0 +1,118 @@
+  RCLCPP_PUBLIC
+  virtual
+  rclcpp::context::Context::SharedPtr
+  get_context() const;
```
If the method is `const` shouldn't it return a const shared pointer? Otherwise the caller can modify the returned value. Maybe we need both: one const method with const return value and one non-const method with a non-const return value?

wjwwood replied:
I don't know. I didn't make this `const` originally (this method was just moved not modified or added) because the shared ptr that is returned is a copy. From what I understand it is not required and not appropriate to return a const of a copy, e.g. `bool get_x() const` and not `const bool get_x() const`. Instead you return a const when you return a reference, e.g. `const MyClass & get_my_class() const` rather than `MyClass & get_my_class() const`.

So if you consider the shared pointer to be a copy, then it doesn't need to be const, i.e. if someone changes the shared ptr that was returned, or calls reset of on it, it would not change the class. However, since it is a pointer, you could modify the thing to which it points and so maybe it should be a `::ConstSharedPtr` as we call it (aka `std::shared_ptr<const Context>`). Getting rid of the shared pointer and just using a raw pointer, it is currently `Context * get_context() const` and maybe it should be `Context const * get_context() const` (reads "the const method get_context returns a pointer to a const Context").

I'm not sure there is a needs to be a const and non const version. I think a non const version is fine, and if the method could be const then that would be better, but not necessary.

So the question is whether or not we should leave it as is, or if we should have it return a `SharedPtr` but not be a const method, or if we should have it return a `ConstSharedPtr`and be a const method.

dirk-thomas replied:
From my point of view if you get a `const Node` then you are not supposed to do anything non-const on the node. But if the class has const methods which expose non-const internals you can call that method and then manipulate the internal state. Therefore const methods should always expose internal state as const.

dirk-thomas commented on rclcpp/include/rclcpp/strategies/allocator_memory_strategy.hpp:
```diff
@@ -52,7 +52,7 @@ class AllocatorMemoryStrategy : public memory_strategy::MemoryStrategy
   using VoidAlloc = typename VoidAllocTraits::allocator_type;
 
-  using WeakNodeVector = std::vector<std::weak_ptr<node::Node>>;
+  // using WeakNodeVector = std::vector<std::weak_ptr<node::Node>>;
```
Can this be removed?

dirk-thomas commented on rclcpp/include/rclcpp/strategies/allocator_memory_strategy.hpp:
```diff
@@ -266,7 +266,7 @@ class AllocatorMemoryStrategy : public memory_strategy::MemoryStrategy
         }
         any_exec->callback_group = group;
-        any_exec->node = get_node_by_group(group, weak_nodes);
+        // any_exec->node = get_node_by_group(group, weak_nodes);
```
Can this be removed?

dirk-thomas commented on rclcpp/include/rclcpp/strategies/allocator_memory_strategy.hpp:
```diff
@@ -300,7 +300,7 @@ class AllocatorMemoryStrategy : public memory_strategy::MemoryStrategy
         any_exec->service = service;
         any_exec->callback_group = group;
-        any_exec->node = get_node_by_group(group, weak_nodes);
+        // any_exec->node = get_node_by_group(group, weak_nodes);
```
Can this be removed?

dirk-thomas commented on rclcpp/include/rclcpp/strategies/allocator_memory_strategy.hpp:
```diff
@@ -333,7 +333,7 @@ class AllocatorMemoryStrategy : public memory_strategy::MemoryStrategy
         any_exec->client = client;
         any_exec->callback_group = group;
-        any_exec->node = get_node_by_group(group, weak_nodes);
+        // any_exec->node = get_node_by_group(group, weak_nodes);
```
Can this be removed?

dirk-thomas commented on rclcpp/include/rclcpp/node_interfaces/node_base.hpp:
```diff
@@ -0,0 +1,118 @@
+  RCLCPP_PUBLIC
+  virtual
+  rcl_node_t *
+  get_rcl_node_handle() const;
```
Same here.

wjwwood replied:
Same thing applies here, despite the fact that it isn't a smart pointer.

dirk-thomas commented on rclcpp/include/rclcpp/node_interfaces/node_base.hpp:
```diff
@@ -0,0 +1,118 @@
+  RCLCPP_PUBLIC
+  virtual
+  std::shared_ptr<rcl_node_t>
+  get_shared_rcl_node_handle() const;
```
Same here.

dirk-thomas commented on rclcpp/include/rclcpp/node_interfaces/node_base.hpp:
```diff
@@ -0,0 +1,118 @@
+  RCLCPP_PUBLIC
+  virtual
+  std::unique_lock<std::recursive_mutex>
+  acquire_notify_guard_condition_lock() const;
```
With the ownership being changed this shouldn't be const?

wjwwood replied:
The returned lock is a temporary and the mutex is mutable, so I'd say no.

dirk-thomas replied:
I did't' know it is a temporary. That makes sense then.

dirk-thomas commented on rclcpp/src/rclcpp/any_executable.cpp:
```diff
@@ -22,8 +22,8 @@ AnyExecutable::AnyExecutable()
-  callback_group(nullptr),
-  node(nullptr)
+  callback_group(nullptr)
+  // node(nullptr)
```
Can this be removed?

dirk-thomas commented on rclcpp/src/rclcpp/node_interfaces/node_base.cpp:
```diff
@@ -0,0 +1,195 @@
+      // Finalize the interrupt guard condition.
+      finalize_notify_guard_condition();
+
+      throw std::runtime_error("failed to interpret ROS_DOMAIN_ID as integral number");
```
While this code block has only been moved this line leaks the memory of `ros_domain_id` on Windows.

wjwwood commented:
I just pushed the graph API refactor and rebased since @Karsten1987's pr was merged.

I'll address the commented out things and other comments once it is in review.

wjwwood commented:
I pushed the `NodeServices` API portion of the refactor and fixed some things added by @Karsten1987's pr which broke due to my changes. Only the `NodeParameters` part is left to go and to decide where to put the "create_timer" functions, either in a `NodeTimers` API part, or in the `NodeBase` part. I'm leaning towards giving timers their own part, mostly because it doesn't seem to fit in the `NodeBase` API very well (most things that use `NodeBaseInterface` don't need timers).

Karsten1987 commented:
+1 for giving timers their own part

gbiggs commented:
I wonder if creating timers, etc. should be done via factory functions that are friends of the base node interface (or node implementation? I forget which is correct for friends). I think that this should allow any object that provides the necessary friend-accessible interface to allow the creation of timers, and would also keep the size of the node object APIs minimal. One of the popular C++ book authors wrote several articles and book chapters about keeping class APIs minimal by making good use of friend and non-friend functions when possible.

I haven't analysed this idea in any way, but I thought it's worth consideration.

gbiggs commented:
Thanks for the diagram and explanations, @wjwwood. It's all stuff I've been hoping for, so I'm really happy to see it happening.

wjwwood commented:
I just pushed Parameters and Timers API's. Now I just need to do some testing and address some existing comments.

wjwwood commented:
Ok, I've addressed or continued discussion on the existing comments. I've also started some CI which will run tonight and I'll check on tomorrow. One thing I'm thinking of doing still is to remove the publisher and subscription factories I created since they don't make as much sense anymore since I've made the templated free functions create_publisher and create_subscription. It's not strictly necessary, but would be easier to read I think, as the factory breaks the flow of the code up considerably.

I put this in review.

wjwwood commented:
Ok, I think I've addressed the constness issues, now there is a tlsf test failing which may be something I broke in the allocator support that I need to look into. Still looking for more reviews in the meantime.

wjwwood commented:
Passing CI:

- Linux:
  - [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=2013)](http://ci.ros2.org/job/ci_linux/2013/)
- OS X:
  - [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=1545)](http://ci.ros2.org/job/ci_osx/1545/)
- Windows:
  - [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=1987)](http://ci.ros2.org/job/ci_windows/1987/)

Karsten1987 commented on rclcpp/include/rclcpp/node.hpp:
```diff
@@ -39,43 +39,25 @@
-};
-
 /// Node is the single point of entry for creating publishers and subscribers.
 class Node : public std::enable_shared_from_this<Node>
```
Does the new architecture allow to create Services inside the node constructor?

wjwwood replied:
Yes, but I haven't tested it.

wjwwood replied:
I'll add a test for this real quick.

wjwwood replied:
I made a test in https://github.com/ros2/system_tests/pull/182 and fixed it in 	f0afcab 😛, good thing you got me to test it.

wjwwood commented on rclcpp/include/rclcpp/create_publisher.hpp:
```diff
@@ -0,0 +1,54 @@
+
+  auto pub = node_topics->create_publisher(
+    topic_name,
+    rclcpp::create_publisher_factory<MessageT, AllocatorT, PublisherT>(allocator),
```
The factory captures all template code into non template functions so they can be interleaved with non template code in the cpp file. As I mentioned before, I think we could remove the factories now in favor of more linear code that lives in the template functions. The trade off being that there is more code in template functions which increase compile time. 

I'd rather do that refactor later though so that we can merge this and let other work build on top in the mean time.

wjwwood commented on rclcpp/include/rclcpp/create_subscription.hpp:
```diff
@@ -0,0 +1,62 @@
+  const rmw_qos_profile_t & qos_profile,
+  rclcpp::callback_group::CallbackGroup::SharedPtr group,
+  bool ignore_local_publications,
+  bool use_intra_process_comms,
```
We need an options structure in c++ for these classes, but I didn't want to tackle that in this pr.

wjwwood replied:
I'm creating an issue for this right now.

wjwwood replied:
https://github.com/ros2/rclcpp/issues/288

wjwwood commented on rclcpp/include/rclcpp/node_interfaces/node_timers.hpp:
```diff
@@ -0,0 +1,60 @@
+class NodeTimers : public NodeTimersInterface
+{
+public:
+  RCLCPP_SMART_PTR_ALIASES_ONLY(NodeTimers)
```
We discussed this before, and while I agree that the thing you ran into was a bad user experience, I don't think it should be addressed in this pr and should perhaps only be changed for classes from which people commonly inherit. Please make a new issue to discuss this.

wjwwood replied:
Can you make an issue for this, so it is tracked?

wjwwood commented on rclcpp/include/rclcpp/create_publisher.hpp:
```diff
@@ -0,0 +1,54 @@
+  node_topics->add_publisher(
+    pub,
+    // this is the callback group, not currently used or exposed to the user for publishers
+    nullptr);
```
I removed the argument in c18b6ab, I'll run CI on Linux only (due to the nature of the change that should be enough).

Karsten1987 commented on rclcpp/include/rclcpp/intra_process_manager_impl.hpp:
```diff
@@ -253,8 +256,20 @@ class IntraProcessManagerImpl : public IntraProcessManagerImplBase
-  using IDTopicMap = std::map<std::string, AllocSet,
-      std::less<std::string>, RebindAlloc<std::pair<const std::string, AllocSet>>>;
+
+  struct strcmp_wrapper : public std::binary_function<const char *, const char *, bool>
```
can we exclude this struct into an external header file? MIght be helpful for other pieces of code.

wjwwood replied:
I don't feel it needs it's own file, it's a pretty trivial class, I would have made it a lambda if I could have. I'd prefer not to separate it, but you can push here with that change if you like.

Karsten1987 commented on rclcpp/include/rclcpp/node.hpp:
```diff
@@ -117,6 +99,14 @@ class Node : public std::enable_shared_from_this<Node>
   rclcpp::callback_group::CallbackGroup::SharedPtr
   create_callback_group(rclcpp::callback_group::CallbackGroupType group_type);
 
+  using CallbackGroup = rclcpp::callback_group::CallbackGroup;
```
can we shift that line before the first usage?

wjwwood replied:
I just removed it and the following line, not sure why I added that there. Might have been a copy-paste issue, thanks for pointing it out. 28a2cfc

Karsten1987 commented on rclcpp/include/rclcpp/node_impl.hpp:
```diff
@@ -73,66 +75,15 @@ Node::create_publisher(
-template<typename MessageT, typename CallbackT, typename Alloc>
+template<typename MessageT, typename CallbackT, typename Alloc, typename SubscriptionT>
 typename rclcpp::subscription::Subscription<MessageT, Alloc>::SharedPtr
 Node::create_subscription(
```
out of consistency, I would also bring that forward function into the `create_subscription.cpp` file.

wjwwood replied:
This is (a) a template function which cannot be put in a cpp file and (b) is a method of the Node class, so it would go in the `node.cpp` file if it could go in a cpp file.

Karsten1987 commented on rclcpp/include/rclcpp/node_impl.hpp:
```diff
@@ -261,30 +140,17 @@ Node::create_subscription(
-typename rclcpp::timer::WallTimer<CallbackType>::SharedPtr
+template<typename DurationT, typename CallbackT>
+typename rclcpp::timer::WallTimer<CallbackT>::SharedPtr
 Node::create_wall_timer(
```
again, just to be consistent, I would also put the `create_wall_timer` into its own file.

wjwwood replied:
Same as above.

Karsten1987 commented on rclcpp/include/rclcpp/node_interfaces/node_base.hpp:
```diff
@@ -0,0 +1,132 @@
+private:
+  RCLCPP_DISABLE_COPY(NodeBase)
+
+  std::string name_;
```
why do we need this? Does it not make sense to hit directly to rcl for getting the name?

wjwwood replied:
No, I updated it in 83b610e. Thanks!

Karsten1987 commented on rclcpp/include/rclcpp/node_interfaces/node_services.hpp:
```diff
@@ -0,0 +1,67 @@
+  RCLCPP_PUBLIC
+  virtual
+  void
+  add_service(
```
I mean why is `add_client` in `NodeServices` and not in `NodeServiceClients` ? Or, if mixed, where is the constructor for `ServiceClient` ?

wjwwood replied:
Clients and Services (Service servers) are two halves of the same part: "services". It is not separate for the same reason that publisher and subscription related functions are together on the `NodeTopics` interface: because they are related. I don't see any need to break up the `NodeServices` interface into `NodeServiceClients` and `NodeServiceServers`, but if you do, please make an argument for that case.

> where is the constructor for ServiceClient?

I don't understand the question, there is no class called `ServiceClient`.

Karsten1987 commented on rclcpp/src/rclcpp/node_interfaces/node_timers.cpp:
```diff
@@ -0,0 +1,47 @@
+
+#include <string>
+
+#include "rclcpp/node_interfaces/node_timers.hpp"
```
shouldn't that include be on first?

wjwwood replied:
I tried that, but I think cpplint was not happy about it for some reason. I'll try again.

wjwwood replied:
I fixed it here and in other places in 23a7728.

Karsten1987 commented on rclcpp/src/rclcpp/node_interfaces/node_topics.cpp:
```diff
@@ -0,0 +1,131 @@
+}
+
+rclcpp::subscription::SubscriptionBase::SharedPtr
+NodeTopics::create_subscription(
```
why is this not in `create_subscription.cpp` ?

wjwwood replied:
Same.

Karsten1987 commented on rclcpp/src/rclcpp/node_interfaces/node_topics.cpp:
```diff
@@ -0,0 +1,131 @@
+{}
+
+rclcpp::publisher::PublisherBase::SharedPtr
+NodeTopics::create_publisher(
```
why is this not in `create_publisher.cpp` ?

wjwwood replied:
Because it is a member of `NodeTopics`?

Karsten1987 commented:
This PR looks good to me. I had a few comments/questions about file ordering and consistency. But no show-stoppers.

wjwwood commented:
Ok, @Karsten1987 I think I addressed your comments, can you check to make sure I did? Also, anyone else that wants to review, please do so now.

I'm starting new CI since I made some significant changes.

wjwwood commented:
The warnings are being addressed by @codebot and I cannot reproduce the OS X test failures on my machine, and I am debugging the Windows test timeouts and I will fix that up after merging this if this pr turns out to have caused it.
